File: graph/src/841_keys_rooms.c
#include <stdlib.h>

void checkGraph(int idx, int** rooms, int* roomsColSize, int* visited){
    visited[idx] = 1;
    for(int i = 0; i < roomsColSize[idx]; i++){
        if(visited[rooms[idx][i]])continue;
        checkGraph(rooms[idx][i], rooms, roomsColSize, visited);
    }
}

int canVisitAllRooms(int** rooms, int roomsSize, int* roomsColSize){
    int* visited = (int*)calloc(roomsSize,sizeof(int));

    checkGraph(0, rooms, roomsColSize, visited);

    for(int i = 0;i < roomsSize;i++){
        if(!visited[i])return 0;
    }
    return 1;
}

File: graph/src/841_keys_rooms.py
class Solution:
    def checkGraph(self,idx, rooms, visited):
        visited[idx] = 1
        for i in range(len(rooms[idx])):
            if visited[rooms[idx][i]]:
                continue
            self.checkGraph(rooms[idx][i], rooms, visited)

    def canVisitAllRooms(self, rooms):
        n = len(rooms)
        visited = [0] * n

        self.checkGraph(0, rooms, visited)
        
        for i in range(n):
            if not(visited[i]):return False;

        return True
